### PR TITLE
Index expression exclusions never trigger "not found"

### PR DESCRIPTION
**Pull request by albertzaharovits**
##### When Security is disabled:

Exclusions of names of resources that do not exist will no longer trigger
the "not found" exception. For example, the expression `*,-foo` currently
throws the "not found" exception if there is no `foo` resource (and the
`ignore_unavailable` request option is `false`, which is the default for
many APIs).
This will no longer be the case. An exclusion will ensure the name is not
included and will never raise any errors.

Note that `foo,a*,-foo` will still trigger a "not found" exception
if `foo` does not exist.

##### When Security is enabled:

Nothing changes. Exclusions of resource names that do not exist never
triggered the "not found" exception when Security is enabled.

### CODE REVIEW

**Comment by albertzaharovits** [comment]
Note to reviewers: [b0b3f56](https://github.com/elastic/elasticsearch/pull/90902/commits/b0b3f5616800d93b08be25bdf0263492e9d985f7) is the effective code change.
The other  changes are embelishments such as replacing the `continue` statements which I found messes with my understanding of the code flow, especially when it comes between the definition and the usage of a variable.

**Review by grcevski — APPROVED** [review]
LGTM! This looks great to me!

**Comment by grcevski** [comment]
The test failure is a known issue.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 1]
```diff
@@ -1263,34 +1271,26 @@ private static void validateAliasOrIndex(String expression) {
             if (expression.charAt(0) == '_') {
                 throw new InvalidIndexNameException(expression, "must not start with '_'.");
             }
+            return expression;
         }
 
-        private static boolean aliasOrIndexExists(Context context, String expression, boolean throwExceptionIfAbsent) {
+        private static RuntimeException aliasOrIndexExists(Context context, String expression) {
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:1277]: It's pretty expensive to raise an exception, especially if you are not using it. Would it be possible to wrap the RuntimeException in a supplier and lazily raise the exception  ?

Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I have pushed [bbcce03](https://github.com/elastic/elasticsearch/pull/90902/commits/bbcce03d9d2ad4f2039f5067d512c2c22ecee05b), I think this addresses your concern.

**Comment by albertzaharovits** [comment]
@jakelandis please take another look.

**Review by jakelandis — APPROVED** [review]
LGTM

**Pull request merged**